Adding `#pragma CPROVER enable "named-check"`

remi-delmas-3000 opened the pull request:
This PR adds the possibility to selectively activate checks :
- on blocks of C code using `#pragma CPROVER enable "named-check"`
- to individual GOTO instructions and expressions using `source_location.add_pragma("enable:named-check")`

It is possible to enable/disable a check, to push a level on the stack and override the check status using another pragma enable/disable pragma.
Attempts to enable and disable a same check at the same stack level are detected and cause a PARSING_ERROR in the C front end and a INVARIANT violation in `goto_check`.

Modifications:
- the collection of allocations on the whole goto program using `goto_checkt::collect_allocation` is now always on
- a `messaget log` attribute is added to `goto_checkt` to be able to log warnings in case a same check is enabled/disabled more than once on a given GOTO instruction
- the `flag_resett` class is extended to detect multiple 'set_flag' operations on a same flag
- the ansi-c lexer/parser is modified to accept `#pragma CPROVER enable "named check"` and flag shadowing
- tests are added in `regression/cbmc/pragma_cprover_enable*` to test the feature

This feature is needed as part of the function and loop contracts verification, to be able to selectively activate checks on  instructions used for assigns-clause checking, in order to guarantee the soundness of the analysis, even if these checks are not globally activated in the current or ulterior `goto-instrument` or `cbmc` passes

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

SaswatPadhi commented on src/analyses/goto_check.cpp:
```diff
@@ -99,6 +103,12 @@ class goto_checkt
+
+  /// when true causes a forced collection of all allocations
+  /// (needed to allow selectively enabling checks per instruction)
+  bool do_collect_allocations;
```
It seems like we are setting this to `true` in the constructor and no other method sets it to `false`, so wouldn't this be _always_ `true`? And if that's the case, can we completely remove this and also remove the `if(!... & !...) return;` check in `collect_allocations`?

Or did I miss a code path where it won't be `true`?

feliperodri replied:
+1

remi-delmas-3000 replied:
this variable was removed and collection is now always on

SaswatPadhi commented on src/analyses/goto_check.cpp:
```diff
@@ -1856,24 +1866,47 @@ optionalt<exprt> goto_checkt::rw_ok_check(exprt expr)
+    {
+      log.warning() << "Found more than one CPROVER enable/disable check "
+                       "pragma on a same goto instruction"
+                    << messaget::eom;
```
I think the error message could be improved a bit. Having "more than one CPROVER enable/disable check" is fine as long as they are _different_ checks.
I think we want to say that we _the same check_ being enabled/disabled multiple times times at a goto instruction.

remi-delmas-3000 replied:
message updated to include the name of the check

SaswatPadhi commented on regression/cbmc/pragma_cprover_enable3/test.desc:
```diff
@@ -0,0 +1,18 @@
+^SIGNAL=0$
+--
+^\[main.pointer_primitives.\d+\] line 17 
+--
```
[nit] missing newline. Also in a few other files.

remi-delmas-3000 replied:
fixed

SaswatPadhi commented on regression/cbmc/pragma_cprover_enable1/main.c:
```diff
@@ -0,0 +1,16 @@
+#pragma CPROVER check enable "signed-overflow"
+  // generate assertions for the following statement
+  x = x + y[1];
+  // pop all annotations
```
```suggestion
  // pop all checks added since the last push
```

[nit] also in other test files

feliperodri replied:
+1

remi-delmas-3000 replied:
comment removed

SaswatPadhi reviewed (commented):
Nice :) Mostly minor comments below.

Can we add at least one regression test to check nesting of both disable and enable pragmas, something like:

```c
  int n;

#pragma CPROVER check push
#pragma CPROVER check enable "signed-overflow"
   // generate overflow checks for the following addition
  int x = n + n;
#pragma CPROVER check push
#pragma CPROVER check disable "signed-overflow"
   // do not generate overflow checks for the following increment operations
  ++n;
  n++;
#pragma CPROVER check pop
   // generate overflow checks for the following addition
  int b = n + n;
#pragma CPROVER check pop
```

feliperodri commented on doc/cprover-manual/properties.md:
```diff
@@ -134,10 +134,13 @@ As all of these checks apply across the entire input program, we may wish to
 use `#pragma CPROVER check disable "<name_of_check>"`, which remains in effect
 until a `#pragma CPROVER check pop` (to re-enable all properties
-disabled before or since the last `#pragma CPROVER check push`) is provided.
+disabled or enabled before or since the last `#pragma CPROVER check push`) is provided. 
```
```
(to re-enable all properties disabled or enabled before or since the last `#pragma CPROVER check push`)
```
This wording seems confusing. Could you re-write it?

remi-delmas-3000 replied:
documentation updated

feliperodri commented on doc/cprover-manual/properties.md:
```diff
@@ -149,6 +152,22 @@ unsigned foo(unsigned x)
+To enable property generation,
+use `#pragma CPROVER check enable "<name_of_check>"`, which remains in effect
+until a `#pragma CPROVER check pop` (to re-enable all properties
+disabled or enabled before or since the last `#pragma CPROVER check push`) is provided.
```
Same as above.

remi-delmas-3000 replied:
documentation updated

feliperodri commented on regression/cbmc/pragma_cprover_enable1/test.desc:
```diff
@@ -0,0 +1,12 @@
+^SIGNAL=0$
+--
+--
+Checks that we can selectively activate checks using pragmas.
```
Add newline.

feliperodri reviewed (approved):
Only minor comments (modulo Saswat's comments).

chrisr-diffblue commented:
Just a general git housekeeping comment - it's generally better to use `git pull --rebase origin develop` (replace `origin` with whatever you've called your remote - `origin` being the default if you didn't specify/change the remote name) to update your branch with the latest `develop` branch, rather than doing a merge of `develop` - it will lead to cleaner commits and cleaner git history.

tautschnig commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -1332,7 +1332,7 @@ void goto_instrument_parse_optionst::instrument_goto_program()
   }
 
   // add generic checks, if needed
-  goto_check(options, goto_model);
```
Use `ui_message_handler` here (should be the same as `log.get_message_handler()`, but is more consistent with the other code over here.

remi-delmas-3000 replied:
corrected

tautschnig commented on src/ansi-c/scanner.l:
```diff
@@ -402,29 +403,33 @@ void ansi_c_scanner_init()
+                  PARSER.pragma_cprover_add_check(check_name, enable);
                   PARSER.set_pragma_cprover();
                 }
-
```
Unnecessary removal of blank line.

tautschnig commented on src/ansi-c/ansi_c_parser.h:
```diff
@@ -143,15 +151,25 @@ class ansi_c_parsert:public parsert
+  /// \brief Adds a check to the CPROVER pragma stack
+  void pragma_cprover_add_check(irep_idt name, bool enabled);
+
+  /// Returns true the same check with different polarity
```
"Returns true iff the same" (missing "iff")

tautschnig commented on src/ansi-c/ansi_c_parser.h:
```diff
@@ -117,20 +123,22 @@ class ansi_c_parsert:public parsert
   ansi_c_id_classt lookup(
     const irep_idt &base_name, // in
-    irep_idt &identifier, // out
+    irep_idt &identifier,      // out
```
Unnecessary whitespace change?

tautschnig commented on src/ansi-c/ansi_c_parser.h:
```diff
@@ -117,20 +123,22 @@ class ansi_c_parsert:public parsert
   {
-    const scopet &current=current_scope();
+    const scopet &current = current_scope();
     scopes.push_back(scopet());
-    scopes.back().prefix=current.prefix+prefix;
+    scopes.back().prefix = current.prefix + prefix;
```
Whitespace change?

tautschnig commented on src/ansi-c/ansi_c_parser.h:
```diff
@@ -107,7 +107,13 @@ class ansi_c_parsert:public parsert
   }
 
-  enum class decl_typet { TAG, MEMBER, PARAMETER, OTHER };
+  enum class decl_typet
+  {
+    TAG,
+    MEMBER,
+    PARAMETER,
+    OTHER
+  };
 
```
Whitespace change?

tautschnig commented on src/ansi-c/ansi_c_parser.h:
```diff
@@ -48,12 +48,12 @@ class ansi_c_parsert:public parsert
-    tag_following=false;
-    asm_block_following=false;
-    parenthesis_counter=0;
+    tag_following = false;
+    asm_block_following = false;
+    parenthesis_counter = 0;
```
Whitespace change?

tautschnig commented on src/ansi-c/ansi_c_parser.h:
```diff
@@ -39,7 +39,7 @@ class ansi_c_parsert:public parsert
   virtual bool parse() override
   {
-    return yyansi_cparse()!=0;
+    return yyansi_cparse() != 0;
```
Whitespace change?

tautschnig commented on src/ansi-c/ansi_c_parser.h:
```diff
@@ -6,21 +6,21 @@ Author: Daniel Kroening, kroening@kroening.com
 int yyansi_cparse();
 
-class ansi_c_parsert:public parsert
+class ansi_c_parsert : public parsert
```
Whitespace change?

tautschnig commented on src/ansi-c/ansi_c_parser.cpp:
```diff
@@ -156,13 +143,12 @@ void ansi_c_parsert::add_declarator(
+    type.id() == ID_struct || type.id() == ID_union || type.id() == ID_c_enum)
     return ansi_c_id_classt::ANSI_C_TAG;
-  else if(type.id()==ID_merged_type)
+  else if(type.id() == ID_merged_type)
```
Whitespace change?

tautschnig commented on src/ansi-c/ansi_c_parser.cpp:
```diff
@@ -156,13 +143,12 @@ void ansi_c_parsert::add_declarator(
-  else if(type.id()==ID_struct ||
-          type.id()==ID_union ||
-          type.id()==ID_c_enum)
+  else if(
+    type.id() == ID_struct || type.id() == ID_union || type.id() == ID_c_enum)
```
Whitespace change?

tautschnig commented on src/ansi-c/ansi_c_parser.cpp:
```diff
@@ -156,13 +143,12 @@ void ansi_c_parsert::add_declarator(
 ansi_c_id_classt ansi_c_parsert::get_class(const typet &type)
 {
-  if(type.id()==ID_typedef)
+  if(type.id() == ID_typedef)
```
Whitespace change?

tautschnig commented on jbmc/src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -812,7 +812,11 @@ void jbmc_parse_optionst::process_goto_function(
+    function.get_goto_function(),
+    ns,
+    options,
+    log.get_message_handler());
```
Use `ui_message_handler`

remi-delmas-3000 replied:
corrected

tautschnig commented on jbmc/src/janalyzer/janalyzer_parse_options.cpp:
```diff
@@ -705,7 +705,11 @@ void janalyzer_parse_optionst::process_goto_function(
+    function.get_goto_function(),
+    ns,
+    options,
+    log.get_message_handler());
```
Use `ui_message_handler`

remi-delmas-3000 replied:
corrected

tautschnig commented on jbmc/src/jdiff/jdiff_parse_options.cpp:
```diff
@@ -278,7 +278,7 @@ bool jdiff_parse_optionst::process_goto_program(
     // add generic checks
     log.status() << "Generic Property Instrumentation" << messaget::eom;
-    goto_check(options, goto_model);
+    goto_check(options, goto_model, log.get_message_handler());
```
Use `ui_message_handler`

remi-delmas-3000 replied:
corrected

tautschnig commented on src/analyses/goto_check.cpp:
```diff
@@ -1892,7 +1904,6 @@ void goto_checkt::goto_check(
     util_make_unique<local_bitvector_analysist>(goto_function, ns);
 
   goto_programt &goto_program=goto_function.body;
-
```
Removal of blank line?

remi-delmas-3000 replied:
corrected

tautschnig commented on src/analyses/goto_check.cpp:
```diff
@@ -1902,32 +1913,64 @@ void goto_checkt::goto_check(
+      else if(d.first == "enable:pointer-primitive-check")
+        seen = flag_resetter.set_flag(enable_pointer_primitive_check, true);
+
+      INVARIANT(
+        !seen,
+        "Found enable and disable pragmas for " + id2string(d.first) +
+          " at \n" + i.source_location().pretty());
```
Shouldn't this invariant just be maintained within `set_flag`? It seems it is checked every time you use `set_flag`.

remi-delmas-3000 replied:
The invariant is now maintained inside `flag_resett`. To be able to generate readable error messages`flag_resett::set_flag` now takes the flag name and instruction as parameters.

tautschnig commented on src/ansi-c/ansi_c_parser.cpp:
```diff
@@ -19,21 +19,18 @@ ansi_c_id_classt ansi_c_parsert::lookup(
-    base_name;
+  const irep_idt scope_name =
+    tag ? "tag-" + id2string(base_name)
+        : label ? "label-" + id2string(base_name) : base_name;
```
Whitespace changes?

tautschnig commented on src/ansi-c/ansi_c_parser.cpp:
```diff
@@ -19,21 +19,18 @@ ansi_c_id_classt ansi_c_parsert::lookup(
 
-  for(scopest::const_reverse_iterator it=scopes.rbegin();
-      it!=scopes.rend();
+  for(scopest::const_reverse_iterator it = scopes.rbegin(); it != scopes.rend();
```
Please review all changes in this file for whitespace-only changes.

tautschnig reviewed (changes requested):
Functionality looks ok, but I think this PR needs some cleanup to remove as much noise as possible.

SaswatPadhi commented on src/ansi-c/ansi_c_parser.h:
```diff
@@ -10,6 +10,7 @@ Author: Daniel Kroening, kroening@kroening.com
 #ifndef CPROVER_ANSI_C_ANSI_C_PARSER_H
 #define CPROVER_ANSI_C_ANSI_C_PARSER_H
 
+#include <map>
```
Do we still need the `#include <set>` below?

SaswatPadhi commented on src/ansi-c/ansi_c_parser.h:
```diff
@@ -143,15 +144,25 @@ class ansi_c_parsert:public parsert
-    }
-  }
+  /// \brief True iff the CPROVER pragma stack is empty
+  bool pragma_cprover_empty();
```
I see why you have introduced these functions although `pragma_cprover_stack` is `public`.

Can we make it `private` / `protected` now? Just so we don't allow messing with `pragma_cprover_stack` directly and use these functions instead?

SaswatPadhi commented on src/ansi-c/ansi_c_parser.cpp:
```diff
@@ -175,3 +175,56 @@ ansi_c_id_classt ansi_c_parsert::get_class(const typet &type)
+
+  for(const auto &pragma : flattened)
+  {
+    std::string check_name = id2string(pragma.first);
```
```suggestion
    auto check_name = id2string(pragma.first);
```

SaswatPadhi commented on src/ansi-c/ansi_c_parser.cpp:
```diff
@@ -175,3 +175,56 @@ ansi_c_id_classt ansi_c_parsert::get_class(const typet &type)
+    std::string check_name = id2string(pragma.first);
+    std::string full_name = pragma.second
+                              ? std::string("enable:") + check_name
+                              : std::string("disable:") + check_name;
```
Since `check_name` gets `std::string` (not `const char *`) type from the `auto` definition above, we can use the overloaded `+` in C++ to simplify this expression:

```suggestion
    auto full_name = (pragma.second ? "enable:" : "disable:") + check_name;
```

remi-delmas-3000 replied:
modif added

SaswatPadhi commented on src/analyses/goto_check.cpp:
```diff
@@ -1902,32 +1917,62 @@ void goto_checkt::goto_check(
+      };
       if(d.first == "disable:bounds-check")
-        flag_resetter.set_flag(enable_bounds_check, false);
+        flag_resetter.set_flag(enable_bounds_check, false, err_msg);
```
I am not sure if I understand why `err_msg` needs to be a lambda function. It seems like it's called immediately (at the `INVARIANT` check) within the `set_flag` call. Can we not just pass a `string` here instead?

remi-delmas-3000 replied:
this lambda is used to defer the creation of the error string until it is actually needed (if ever).

SaswatPadhi commented on src/analyses/goto_check.cpp:
```diff
@@ -1995,7 +2040,11 @@ void goto_checkt::goto_check(
+          return "no_enum_check flag disabled and enabled at \n" +
+                 i.source_location().pretty();
+        };
+        no_enum_check_flag_resetter.set_flag(no_enum_check, true, err_msg);
```
Same comment as above.
Can we not just pass a `string` here instead?

remi-delmas-3000 replied:
This lambda allows to defer the creation of the string until it is needed

remi-delmas-3000 replied:
now passing flag name and instructions to set_flag

SaswatPadhi commented on src/analyses/goto_check.cpp:
```diff
@@ -1853,27 +1851,44 @@ optionalt<exprt> goto_checkt::rw_ok_check(exprt expr)
 
-  /// Restore the values of all flags that have been modified via `set_flag`.
+  /// \brief True iff flag has already been seen by \ref set_flag
+  bool seen(bool &flag)
```
Hmm, does this function require a mutable reference? I can see why `set_flag` does, but I don't think `seen` does? How about:

```suggestion
  bool seen(const bool &flag)
```

SaswatPadhi commented on src/analyses/goto_check.cpp:
```diff
@@ -1853,27 +1851,44 @@ optionalt<exprt> goto_checkt::rw_ok_check(exprt expr)
+  /// \brief True iff flag has already been seen by \ref set_flag
+  bool seen(bool &flag)
+  {
+    return seen_flags.find(&flag) != seen_flags.end();
```
Hmm, we seem to be taking the address of `flag` here. So if we have two `flag`s with the same content but different addresses then this would break?

Do we have some mechanism to guard against it? Or if there's some other reason as to why this is a non-issue, we should mention that in the comments above this function.

remi-delmas-3000 replied:
the class `flag_resett` is file-local so as long as the discipline is respected in this file it should be ok.

SaswatPadhi reviewed (changes requested):
A few more comments; nothing major

tautschnig commented on src/analyses/goto_check.cpp:
```diff
@@ -1851,30 +1849,54 @@ optionalt<exprt> goto_checkt::rw_ok_check(exprt expr)
     if(flag != new_value)
     {
-      flags_to_reset.emplace_back(&flag, flag);
+      flags_to_reset.insert(std::pair<bool *, bool>(&flag, flag));
```
I might be wrong, but I believe you can shorten this to `flags_to_reset.insert({&flag, flag});` with C++11.

SaswatPadhi replied:
Or just `flags_to_reset.emplace(&flag, flag);` to avoid creating a temporary pair.

tautschnig commented on src/analyses/goto_check.cpp:
```diff
@@ -1851,30 +1849,54 @@ optionalt<exprt> goto_checkt::rw_ok_check(exprt expr)
+      flags_to_reset.insert(std::pair<bool *, bool>(&flag, flag));
       flag = new_value;
     }
+    return;
```
Seems redundant?

remi-delmas-3000 replied:
fixed

tautschnig commented on src/analyses/goto_check.cpp:
```diff
@@ -1851,30 +1849,54 @@ optionalt<exprt> goto_checkt::rw_ok_check(exprt expr)
 private:
-  std::list<std::pair<bool *, bool>> flags_to_reset;
+  /// \brief True iff flag has already been seen by \ref set_flag
+  bool seen(bool *flag)
```
Not that it matters a whole lot, but the parameter type should be `const bool *` to make clear that no modification will occur. Also, the method should be marked `const`.

remi-delmas-3000 replied:
method removed

tautschnig commented on src/analyses/goto_check.cpp:
```diff
@@ -1898,36 +1920,62 @@ void goto_checkt::goto_check(
     goto_programt::instructiont &i=*it;
 
-    flag_resett flag_resetter;
+    flag_resett resetter;
```
Given that every call requires the instruction to be passed (and the instruction is always the same), should the constructor already take that?

remi-delmas-3000 replied:
fixed

tautschnig commented on src/ansi-c/ansi_c_parser.cpp:
```diff
@@ -175,3 +175,55 @@ ansi_c_id_classt ansi_c_parsert::get_class(const typet &type)
+
+void ansi_c_parsert::pragma_cprover_push()
+{
+  std::map<irep_idt, bool> empty;
+  pragma_cprover_stack.push_back(std::move(empty));
```
I'm wondering whether `pragma_cprover_stack.emplace_back({});` would actually work (and be more concise)?

remi-delmas-3000 replied:
seems to work with `pragma_cprover_stack.emplace_back()`

tautschnig commented on src/ansi-c/ansi_c_parser.cpp:
```diff
@@ -175,3 +175,55 @@ ansi_c_id_classt ansi_c_parsert::get_class(const typet &type)
+  pragma_cprover_stack.pop_back();
+}
+
+void ansi_c_parsert::pragma_cprover_add_check(irep_idt name, bool enabled)
```
Use `const irep_idt &`

remi-delmas-3000 replied:
fixed

tautschnig commented on src/ansi-c/ansi_c_parser.cpp:
```diff
@@ -175,3 +175,55 @@ ansi_c_id_classt ansi_c_parsert::get_class(const typet &type)
+  pragma_cprover_stack.back()[name] = enabled;
+}
+
+bool ansi_c_parsert::pragma_cprover_clash(irep_idt name, bool enabled)
```
`const irep_idt &`

remi-delmas-3000 replied:
fixed

SaswatPadhi commented on src/ansi-c/ansi_c_parser.cpp:
```diff
@@ -175,3 +175,55 @@ ansi_c_id_classt ansi_c_parsert::get_class(const typet &type)
+
+  for(const auto &pragma : flattened)
+  {
+    std::string check_name = id2string(pragma.first);
+    std::string full_name =
+      (pragma.second ? "enable:" : "disable:") + check_name;
```
```suggestion
    auto check_name = id2string(pragma.first);
    auto full_name =
      (pragma.second ? "enable:" : "disable:") + check_name;
```

remi-delmas-3000 replied:
`scanner.l ` uses `std::string` everywhere so I'll keep this as is for consistency

SaswatPadhi replied:
Yes, but not sure why it has to be kept this way here though ... Using `auto` is not only more succinct but also makes it future proof (disconnects this from the underlying type, which is base `std::string` today)

remi-delmas-3000 replied:
this would only be patching two lines out of several dozen that could use `auto`  the whole lexer would need refactoring to be effectively future proof.

SaswatPadhi replied:
Up to you 🙂

SaswatPadhi reviewed (approved):
Approving modulo Michael's remaining comments and one minor suggestion:

feliperodri commented:
@peterschrammel @martin-cs could you take a look at this PR? It's missing code owner approval.

martin-cs commented:
@feliperodri Apologies for the lag; I missed this one.

martin-cs commented:
This would be easier to review if it were split up into multiple commits.  Parsing, implementation, testing would seem a reasonable start.

Commit:
Adding support for check enable pragmas in the ansi-c lexer and parser

Adding enable and disable pragmas for a same check at a same pragma stack level triggers an error.
Enable and disable pragmas for a same check in different stack levels is allowed, and shadowing is managed when flattening the stack.

Commit:
Adds a message_handlert to goto_check functions and goto_checkt.

No change in behaviour expected. No need for specific tests.

Commit:
collect_allocations is now always on.

Needed so that the data required to express pointer checks
is always available to process a enable pragmas.

Commit:
Adding extra logic to flat_resett to detect and trigger errors on double updates with opposite values.

Commit:
Adds processing of enable pragmas with flag_resett

Commit:
Regression tests for enable pragmas

Commit:
Remove unused import

Commit:
Updated documentation for CPROVER check enable and disable pragmas.

remi-delmas-3000 commented:
@martin-cs @feliperodri @tautschnig I reworked the commit history it should be good for the final review now

Commit:
Removing unused import

martin-cs commented:
Thank you @remi-delmas-3000 it makes reviewing so much easier!
